docs/proposals/scheduling/20231205-dynamic-reservation-provisioning.md

```diff
@@ -0,0 +1,215 @@
+5. Completion of the Binding Process: After the scheduler receives notification of a new Reservation CRD instance, it replaces the virtual object with the actual reservation, adjusts the reservation allocable resources by an amount equal to the sum of the requests of the associated Pods, and continues to bind the Pod to the appropriate node.
+
+### Fault tolerance
+
+Once provisioner exception prevents reservation from being created, fault tolerance needs to be implemented:
+
+- The scheduler should have a default wait timeout, similar to the mechanism in Dynamic Volume Provisioning. If a Reservation is not created after waiting for 10 minutes, the process should be rolled back, and the Pod should be re-queued for another scheduling attempt.
+- Consider retaining the virtual reservation object; it could be reused in the next Pod scheduling attempt. It is recommended to keep it at first. Once the provisioner recovers from any exceptions, it can attempt to create a new Reservation object that can replace the virtual reservation object in memory.
+- In terms of the resource delivery success rate, if the creation of a Reservation fails, the Pod's scheduling could still be completed, and the virtual reservation object should be deleted.
+
+### API
+
+#### Reservation Template Annotation
+
+Annotation key: `scheduling.koordinator.sh/reservation-template`, defined as follows：
+
+```go
+type ReservationTemplate struct {
+	Owners          []schedulingv1alpha1.ReservationOwner  `json:"owners"`
+	RecommendedSize corev1.ResourceList                    `json:"recommendedSize"`
+	RoundUp         corev1.ResourceList                    `json:"roundup"`
+}
+```
+
+The `recommendedSize` represents the desired maximum specification for a Reservation, but it is not mandatory. When a Pod is scheduled and has the `scheduling.koordinator.sh/reservation-template` annotation, the scheduler calculates the maximum between the Pod's requests and recommendedSize, using this maximum value for scheduling. If the Pod's requests are less than recommendedSize and the remaining resources in cluster nodes are insufficient to meet recommendedSize, the scheduler will attempt to schedule the Pod using the Pod's requests, aiming to schedule the Pod as efficiently as possible.
+
+For example:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  annotations:
+    scheduling.koordinator.sh/reservation-template: |
+      {
+        "owners": [
+          {
+            "labelSelector": {
+              "matchLabels": {
+                "tenancy-id": "1612285282502324"
+              }
+            }
+          }
+        ],
+        "recommendedSize": {
+          "cpu": "4",
+          "memory": "8Gi"
+        },
+        "roundup": {
```

Comment:
I think we can keep the strategy simple first and don’t need to add roundup for the time being.